File: 15.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = (int)(nums.size());
        vector<vector<int>> v;
        sort(nums.begin(), nums.end());
        
        for(int i = 0; i < n-2; i++){
            if(i > 0 && nums[i] == nums[i-1]) continue;
            for(int l = i + 1, r = n - 1; l < r;){
                long long sum3 = (long long)nums[i] + nums[l] + nums[r];
                if(sum3 == 0){
                    v.push_back({nums[i], nums[l], nums[r]});

                    while(l < r && nums[l] == nums[l+1]) l++;
                    while(r > l && nums[r] == nums[r-1]) r--;

                    l++;
                    r--;
                }

                if(sum3 < 0) l++;
                else if(sum3 > 0) r--;
            }
        }
        return v;
    }
};

File: 15.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = []
        nums.sort()

        for i in range(n-2):
            if i > 0 and nums[i] == nums[i-1]: continue
            l, r = i + 1, n - 1
            while l < r:
                sum3 = nums[i] + nums[l] + nums[r]
                if sum3 == 0:
                    ans.append([nums[i], nums[l], nums[r]])

                    while(l < r and nums[l] == nums[l+1]): l += 1
                    while(r > l and nums[r] == nums[r-1]): r -= 1

                    l += 1
                    r -= 1

                if sum3 < 0:
                    l += 1
                elif sum3 > 0:
                    r -= 1

        return ans
